tokio/src/time/interval.rs

```diff
@@ -90,39 +99,282 @@ pub fn interval(period: Duration) -> Interval {
+///
+/// # Examples
+///
+/// ```
+/// use tokio::time::{
+///     interval_with_missed_tick_behavior,
+///     sleep,
+///     Duration,
+///     MissedTickBehavior,
+/// };
+///
+/// #[tokio::main]
+/// async fn main() {
+///     let mut interval = interval_with_missed_tick_behavior(
+///         Duration::from_millis(10),
+///         MissedTickBehavior::Skip,
+///     );
+///
+///     interval.tick().await;                  // ticks immediately
+///     sleep(Duration::from_millis(25)).await; // simulating executor getting blocked
+///     interval.tick().await;                  // ticks immediately, yielding an `Instant` 10ms after the starting tick
+///     interval.tick().await;                  // ticks after 5ms, yielding an `Instant` 30ms after the starting tick
+///
+///     // approximately 30ms have elapsed.
+/// }
+/// ```
+pub fn interval_with_missed_tick_behavior(
+    period: Duration,
+    missed_tick_behavior: MissedTickBehavior,
+) -> Interval {
+    assert!(period > Duration::new(0, 0), "`period` must be non-zero.");
+
+    interval_with_missed_tick_behavior_at(Instant::now(), period, missed_tick_behavior)
+}
+
+/// Creates new [`Interval`] with a specific [`MissedTickBehavior`] that yields
+/// with interval of `period`. The first tick completes at `start`.
+///
+/// An interval will tick indefinitely. At any time, the [`Interval`] value can
+/// be dropped. This cancels the interval.
+///
+/// # Panics
+///
+/// This function panics if `period` is zero.
+///
+/// # Examples
+///
+/// ```
+/// use tokio::time::{
+///     interval_with_missed_tick_behavior_at,
+///     sleep,
+///     Duration,
+///     Instant,
+///     MissedTickBehavior
+/// };
+///
+/// #[tokio::main]
+/// async fn main() {
+///     let start = Instant::now() + Duration::from_millis(50);
+///     let mut interval = interval_with_missed_tick_behavior_at(
+///         start,
+///         Duration::from_millis(10),
+///         MissedTickBehavior::Skip,
+///     );
+///
+///     interval.tick().await;                  // ticks after 50ms
+///     sleep(Duration::from_millis(25)).await; // simulating executor getting blocked
+///     interval.tick().await;                  // ticks immediately, yielding an `Instant` 10ms after the starting tick
+///     interval.tick().await;                  // ticks after 5ms, yielding an `Instant` 30ms after the starting tick
+///
+///     // approximately 30ms have elapsed.
+/// }
+/// ```
+pub fn interval_with_missed_tick_behavior_at(
+    start: Instant,
+    period: Duration,
+    missed_tick_behavior: MissedTickBehavior,
+) -> Interval {
+    assert!(period > Duration::new(0, 0), "`period` must be non-zero.");
+
     Interval {
         delay: Box::pin(sleep_until(start)),
         period,
+        missed_tick_behavior,
+    }
+}
+
+/// Defines the behavior of an [`Interval`] when it misses a tick.
+///
+/// Occasionally, the executor may be blocked, and as a result, [`Interval`]
```

Comment:
The executor being blocked is typically not the reason that ticks are missed. Ticks can be missed because whatever the async task was doing took a long time, and that operation may very well be non-blocking. For example, consider this case:
```Rust
#[tokio::main]
async fn main() {
    let mut interval = time::interval(time::Duration::from_secs(2));
    for _i in 0..5 {
        interval.tick().await;
        task_that_takes_one_to_three_seconds().await;
    }
}
```
Here the reason for ticks being missed is that `task_that_takes_one_to_three_seconds` took more than two seconds.

Reply:
Okay. Are there any other conditions that I should bring up where a tick could be missed, besides a task being (asynchronously) blocked or the executor (rarely) getting blocked?

Reply:
No. Generally the way you should explain it is that the `MissedTickBehavior` controls what happens if someone spends too long without calling `tick()`. You don't have to mention the whole blocked-runtime thing.